### ISSUE TITLE
Spring Cloud Bus mocking?

### ISSUE THREAD

**Issue by alecl** [fridujo/rabbitmq-mock#25]
Can this be configured to work through the Spring Cloud Bus abstraction to Rabbit?

**Comment by ledoyen** [comment]
TL;DR yes it should.

I am not familiar with **spring-cloud-bus** however, it seems that:
- **spring-cloud-bus** make use of RabbitMQ through **spring-cloud-stream-binder-rabbit**
- this particular library binds the **spring-rabbit** `ConnectionFactory` with the  *auto configuration `RabbitMessageChannelBinderConfiguration`

So in theory, in you just have to override the bean definition of `org.springframework.amqp.rabbit.connection.ConnectionFactory` to make it work (like any other Spring app using **rabbitmq-mock** for integration and/or feature testing).

However, it also may need extrawork as the *auto configuration* pops an `RabbitMessageChannelBinder` which creates a `org.springframework.amqp.rabbit.connection.LocalizedQueueConnectionFactory`, which creates it's own `ConnectionFactory` in the protected method `org.springframework.amqp.rabbit.connection.LocalizedQueueConnectionFactory#createConnectionFactory`.

It may also be interresting for you to look into **spring-cloud-stream-test-support**, to avoid relying on a specific messaging system, which seems to be the point of **spring-cloud-stream** `org.springframework.cloud.stream.binder.AbstractMessageChannelBinder`.

**Comment by ledoyen** [comment]
@alecl I have provided a sample to show how **rabbitmq-mock** can be used to mock underlying **spring-cloud-bus** broker, see #26.

**rabbitmq-mock** do the job, the tricky part of this exercice was to mock the `RestTemplate` used by `ConfigServicePropertySourceLocator` as it is used very early in the context lifecycle. This had to be done using **initializers** (`ApplicationContextInitializer`).

Hope this will help you !

Let me know if you think of any improvement, it is the first time I deal with **spring-cloud-bus**, so many things may have eluded me.
